internal/generator/ecc/template/sswu-fp.go.tmpl

```diff
@@ -0,0 +1,397 @@
+    tv2.Mul(&tv2, u)
+    tv1.Mul(&tv1, &tv2)
+
+    var c1 big.Int
+    c1.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 0) ) }} })
+    var y1 {{.CoordType}}
+    y1.Exp(tv1, &c1)
+    y1.Mul(&y1, &tv2)
+    tv1.Mul(&y1, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 1) }} })
+    tv2.Square(&tv1)
+
+    //Line 10 in std doc
+    tv2.Mul(&tv2, v)
+
+    if tv2 == *u {
+        y1 = tv1
+    }
+
+    tv2.Square(&y1)
+    tv2.Mul(&tv2, v)
+
+    //Line 15
+    isQr := tv2 == *u
+    var y2 {{.CoordType}}
+    y2.Mul(&y1, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 2) }} } )
+    tv1.Mul(&y2, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 1) }} })
+    tv2.Square(&tv1)
+    tv2.Mul(&tv2, v)
+
+    var tv3 {{.CoordType}}
+    //Line 20
+    // Z = {{.Z}}
+    mulByZ(&tv3, u)
+    if tv2 == tv3 {
+        y2 = tv1
+    }
+    if isQr {
+        *z = y1
+    } else {
+        *z = y2
+    }
+    
+    return isQr
+}
+
+{{ end }}
+
+{{ if eq (mod .FieldSizeMod256 8) 1 }}
+// sqrtRatio computes the square root of u/v and returns true if u/v was indeed a quadratic residue
+// if not, we get sqrt(Z * u / v). Recall that Z is non-residue
+// The return value is undefined for u = 0
+// Taken from https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/13/ F.2.1.1. for any field
+// The main idea is that since the computation of the square root involves taking large powers of u/v, the inversion of v can be avoided
+func sqrtRatio(z *{{.CoordType}}, u *{{.CoordType}}, v *{{.CoordType}}) bool {
+   tv1 := {{.CoordType}} { {{words64 (index .SqrtRatioParams 5) }} }
+
+   var exp big.Int
+   exp.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 3) ) }} })
+   var tv2, tv3, tv4, tv5 {{.CoordType}}
+   tv2.Exp(*v, &exp)
+   tv3.Mul(&tv2, &tv2)
+   tv3.Mul(&tv3, v)
+
+   // line 5
+   tv5.Mul(u, &tv3)
+
+   exp.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 2) ) }} })
+   tv5.Exp(tv5, &exp)
+   tv5.Mul(&tv5, &tv2)
+   tv2.Mul(&tv5, v)
+   tv3.Mul(&tv5, u)
+
+   // line 10
+   tv4.Mul(&tv3, &tv2)
+   exp.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 4) ) }} })
+   tv5.Exp(tv4, &exp)
+
+   isQr := tv5.IsOne()
+
+   tv2.Mul(&tv3, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 6) }} })
+   tv5.Mul(&tv4, &tv1)
+
+   // line 15
+
+   if !isQr {
+      tv3, tv4 = tv2, tv5
+   }
+
+   exp.Lsh( big.NewInt(1), {{ index .SqrtRatioParams 0 }} - 2)
+   for i := {{ index .SqrtRatioParams 0 }}; i >= 2; i -- {
+      //line 20
+      tv5.Exp(tv4, &exp)
+      e1 := tv5.IsOne()
+
+      tv2.Mul(&tv3, &tv1)
+      tv1.Mul(&tv1, &tv1)
+      tv5.Mul(&tv4, &tv1)
+
+      if !e1 {
+         tv3, tv4 = tv2, tv5
+      }
+
+      exp.Rsh(&exp,1)
+   }
+
+   *z = tv3
+   return isQr
+}
+{{ end }}
+
+//TODO: Use addchain
+//TODO: Might duplicate functionality from mulByConst functions
+// mulByZ multiplies x by {{.Z}} and stores the result in z
+func mulByZ(z *{{.CoordType}}, x *{{.CoordType}}) {
```

Comment:
unless Z is often small on our curves, I suspect the perf diff with the optimized Mul operation are negligible . 
I would kill this func and use Mul directly.

Reply:
It _is_ small, but we can't use the same trick (or a simple-enough-to-be-worth-it variant) on extensions.